### ISSUE TITLE
Pressing enter moves the last character to a new line when typing Hangul (Korean)

### ISSUE THREAD

**Issue by bjrambo** [ckeditor/ckeditor5#777]
## 🐞 Is this a bug report or feature request? (choose one)

- Bug report

## 💻 Version of CKEditor

https://ckeditor5.github.io/ <- here editor

## 📋 Steps to reproduce

1. Korean typing, (Window Korea Keyboard) if you want typing "한글입니다-> gksrmfdlqslek"
2. Type "한글입니다" and press Enter key immediately.

You can check the symptoms in the image below.

## 📃 Other details that might be useful
![ckeditor5_1](https://user-images.githubusercontent.com/4381756/35206673-b0cc3114-ff81-11e7-9616-250c616d24fe.gif)

**Comment by bjrambo** [comment]
I use a chrome.

**Comment by oleq** [comment]
Possibly related to https://github.com/ckeditor/ckeditor5/issues/777. @Reinmar?

**Comment by Reinmar** [comment]
#777 is definitely related to #777 :D But, it may also be related to #748.

@bjrambo, could you check whether #748 describes the bug that you encounter as well? Because you didn't state what happens and what's the expected result and I can see on your screenshot that #748 occurs there as well, but perhaps you meant also some other issues,

**Comment by bjrambo** [comment]
@Reinmar Seems to be related but different. Japanese has problem at the beginning of line. Korean has problem at the end of line.

**Comment by Reinmar** [comment]
@Mgsy, could you take a look on this?

**Comment by Mgsy** [comment]
I can confirm this issue, but the problem is not related to the end of the line. As you can see on the screencast, the selection is always behind inserted character (except character at the beginning of the block) and that's why one character moves downwards after pressing <kbd>Enter</kbd>.

![bug_cke5](https://user-images.githubusercontent.com/20988892/35566779-e7c835a4-05c2-11e8-8e68-7935c94a640d.gif)

**Comment by Mgsy** [comment]
> I assume that in a plain contenteditable or in CKEditor 4 the selection stays after the underlined part of the content?

I've made an update to my previous comment. In a plain contenteditable the selection stays behind inserted character, but after pressing <kbd>Enter</kbd> it behaves properly.

> Also, does a similar issue occur if you start typing at the end of a non-empty line?

Yes, behaviour is the same.

This bug isn't reproducible in Firefox and Edge, I guess because of different typing method:

![bug_cke5](https://user-images.githubusercontent.com/20988892/35567640-01fd03b6-05c6-11e8-99d7-43e1c067b88e.gif)

**Comment by f1ames** [comment]
I have analysed this issue in a little more detail. I suspected it might be caused by the `renderer` and the fact that it rerenders selection during composition which may break it.

However, it seems the cause is totally different. First of all, the fact that caret is before the last character when composing in Hangul (Korean) is already [reported as a bug in a Chrome bugtracker](https://bugs.chromium.org/p/chromium/issues/detail?id=653160). However as it works correctly in plain _contenteditable_ (even though selection is before the last character - enter does not move it to the new line) it means that CKEditor behaves different in some way and breaks it.

The different behaviour is caused by our custom <kbd>enter</kbd> handling ([Enter feature](https://github.com/ckeditor/ckeditor5-enter)) which relays on the current selection when inserting new lines (in short). Now, when typing regular text like:

```
<p>Foo Ba{}r</p>
```

and when composing Hangul:

```
<p>Foo ㅎㅇ{}ㄷ</p>
```

the native selection (and its ranges) is exactly the same showing that selection is before the last character. So relying on the native selection we are unable to tell that for Hangul the <kbd>enter</kbd> behaviour should be different.

This means that enter behaviour is inconsistent and I have already [reported it to a Chrome team](https://bugs.chromium.org/p/chromium/issues/detail?id=849263). TBH, I didn't find any other way to distinguish the Hangul case from other cases so I'm not sure if there is any reasonable way to fix it for now🤔

**Comment by Clouds101** [comment]
@f1ames 
It does not seem to be a problem with Chrome. There is no problem in ckeditor4.
Please check the image on #1117

**Comment by Reinmar** [comment]
> There is no problem in ckeditor4.

CKEditor 4 works completely differently than CKEditor 5 so this Chrome's bug isn't triggered there.

**Comment by startong21** [comment]
Is there any idea?
The Chrome team seems not to fix this problem and it takes long time already.
We use many "Enter key" in sentences globally and I want to use CKEditor5 in developing Korean sites.

**Comment by Clouds101** [comment]
@startong21 
When entering characters in the editor, if the character you are typing is "enter", you tried to enter "enter" after moving to the right if the previous character was Korean. However, unlike version 4 of ckeditor, I stopped making corrections.

Please note that below is not a complete source code.

var ck5LastHan = false;
editor.editing.view.document.on( 'keydown', ( evt, data ) => {
	var ck5pass = getKeyStatus(data.keyCode);
	if(!ck5pass){
		ck5LastHan = true;
	}else{
		ck5LastHan = false;
	}
} );

ClassicEditor
.create( document.querySelector( '#editor' ) )
.then( editor => {
 editor.keystrokes.set( 'Enter', (data, stop ) => {
	if(ck5LastHan)
	{
	  //give up
	}
}, { priority: 'high' }  ); 

// korean char detect
function getKeyStatus(char_ASCII){
	 if (char_ASCII >= 48 && char_ASCII <= 57 )
	return true;
	else if ((char_ASCII>=65 && char_ASCII<=90) || (char_ASCII>=97 && char_ASCII<=122))
		return true;
	else if ((char_ASCII>=33 && char_ASCII<=47) || (char_ASCII>=58 && char_ASCII<=64) 
	   || (char_ASCII>=91 && char_ASCII<=96) || (char_ASCII>=123 && char_ASCII<=126))
	return true;
	else if (char_ASCII>=188 && char_ASCII<=190) 
	return true;
	else if (char_ASCII>=8 && char_ASCII<=1) 
	return true;

 return false;
}


// You will also need to add the source code above that works only with the Chrome browser.

**Comment by startong21** [comment]
Thank you for your reply!
I've added your code but I've got a console error message - Uncaught TypeError: Cannot read property 'view' of undefined.
So I wrote a simple code using your code in https://jsfiddle.net/2u9nqsvo/2/
Please check this code. :)

**Comment by f1ames** [comment]
Hello @startong21,
I have checked the [849263 issue](https://bugs.chromium.org/p/chromium/issues/detail?id=849263) on the latest Chrome (`71.0.3578.80`) on Windows and it still occurs there.

Please keep in mind that when checking this issue it is the matter of what is the state/ranges of native selection so it can't be simply checked by typing text, pressing enter and checking if last character was moved to the next line. This will work in native text fields and that's also what [849263 issue](https://bugs.chromium.org/p/chromium/issues/detail?id=849263) issue is about. You may take a look at the above comment https://github.com/ckeditor/ckeditor5/issues/777#issuecomment-394639005 for more technical explanation.

**Comment by startong21** [comment]
Hello @f1ames,
I don't have many knowledge about input method and contenteditable so I seem to talk about what it isn't easy or reasonable.

When I checked Chrome bugtracker, [653160 issue](https://bugs.chromium.org/p/chromium/issues/detail?id=653160) looks just caret position issue.
It means it's not functional but just design issue.
In other words, when Korean people type Korean, caret position looks strange (because the caret position is on the left and not right. some Korean says that the caret position should be on the right.), but it is just design issue. And functionally it is no problem.
But this is just Chrome style and other editors in Windows use the block design style, and you can see this in [653160 issue comment 5](https://bugs.chromium.org/p/chromium/issues/detail?id=653160#c5).

I'm not sure how I write the source code for this but how about to solve this problem like the following?
When you type the Enter key typing Korean, you move the caret to the right one step and to new line.
And you may refer to a @Clouds101 comment for Korean character ASCII codes.
I think this can solve almost all cases of typing Korean and enter key problems.

**Comment by f1ames** [comment]
@startong21 The issue is not a caret style/position itself but the inconsistency between caret position / native selection and the way <kbd>Enter</kbd> behaves in this cases. As described in [#653160](https://bugs.chromium.org/p/chromium/issues/detail?id=653160) and https://github.com/ckeditor/ckeditor5/issues/777#issuecomment-394639005 above:

> When typing regular text like:
> 
> ```
> <p>Foo Ba{}r</p>
> ```
> 
> and when composing Hangul:
> 
> ```
> <p>Foo ㅎㅇ{}ㄷ</p>
> ```
> 
> the native selection (and its ranges) is exactly the same showing that selection is before the last character. So relying on the native selection we are unable to tell that for Hangul the <kbd>enter</kbd> behaviour should be different.
> 
> This means that enter behaviour is inconsistent and I have already reported it to a Chrome team.

---

@Clouds101 the solution may work but it will need some adjustments (still it is kind of hackish workaround). Imagine for example that someone typed Korean, then changed selection (moved caret somewhere) and then pressed <kbd>Enter</kbd>. So you cannot reasonably rely on only checking what was the last character typed when <kbd>Enter</kbd> is pressed. What could be done instead:

1. Listen on `keydown` and check if Korean character was inserted (store the result so other listeners may check it).
1. If any action like selection change, new content insertion and so on happens, reset the stored result.
1. If <kbd>Enter</kbd> was pressed check stored result, if it was Korean and there is a Korean character after selection, new line should be inserted after this character.

In other words the special enter handler should only kick in when <kbd>Enter</kbd> is pressed directly (directly meaning that there were no other actions in-between) after typing Korean. Still I'm not 100% sure if you can reasonably handle all cases this way without breaking anything.

**Comment by Clouds101** [comment]
This issue is fixed in Chrome version 75 or later.

**Comment by f1ames** [comment]
> This issue is fixed in Chrome version 75 or later.

It's what the upstream comment also says - https://bugs.chromium.org/p/chromium/issues/detail?id=653160#c18. @Mgsy could you check if it solved the issue in the editor?

**Comment by Mgsy** [comment]
This issue has been fixed.
